File: 1/2020.10.13/322.Coin Change/java/solution.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount < 0) {
            return 0;
        }
        return helper(coins, amount, new int[amount]);
    }

    private int helper(int[] coins, int rem, int[] amount) {
        if (rem < 0) {
            return -1;
        }
        if (rem == 0) {
            return 0;
        }
        if (amount[rem-1] != 0) {
            return amount[rem-1];
        }
        int min = Integer.MAX_VALUE;
        for (int coin : coins) {
            int res = helper(coins, rem-coin, amount);
            if (res >= 0 && res < min) {
                min = res + 1;
            }
        }
        amount[rem-1] = min == Integer.MAX_VALUE ? -1 : min;
        return amount[rem-1];
    }

    public int coinChange(int[] coins, int amount) {
        if (amount < 1) {
            return 0;
        }
        int[] dp = new int[amount + 1];
        for (int i = 1; i <= amount; i++) {
            int min = -1;
            for (int coin : coins) {
                if (i >= coin && dp[i-coin] != -1) {
                    int temp = dp[i-coin] + 1;
                    min = min < 0 ? temp : Math.min(temp, min);
                }
            }
            dp[i] = min;
        }
        return dp[amount];
    }
}

File: 1/2020.10.13/322.Coin Change/python/solution.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # dp
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0
        for coin in coins:
            for i in range(coin, amount+1):
                dp[i] = min(dp[i], dp[i-coin] + 1)
        return [dp[amount], -1][dp[amount] == float('inf')]

    def coinChange(self, coins: List[int], amount: int) -> int:
        @functools.lru_cache(amount)
        def dp(rem):
            if rem < 0:
                return -1
            if rem == 0:
                return 0
            mini = float('inf')
            for coin in coins:
                res = dp(rem - coin)
                if res >= 0 and res < mini:
                    mini = res + 1
            return mini if mini != float('inf') else -1
        return dp(amount)
